nmosregistration/couchbase_backend.py

```diff
@@ -0,0 +1,202 @@
+import string
+from flask import make_response
+
+legacy_key_table = {
+    '@_apiversion': 'api_version'
+}
+
+def _legacy_key_lookup(key):
+    try:
+        return legacy_key_table[key]
+    except KeyError:
+        # TODO: More proper handling of any additional keys starting '@_'?
+        #       - May just be sufficient to keep table and return otherwise, need to see other cases.
+        return key
+
+class CouchbaseInterface(object):
+    def __init__(self, cluster_address, username, password, bucket, *args, **kwargs):
+        self.cluster = couchbase.cluster.Cluster('couchbase://{}'.format(','.join(cluster_address)))
+        auth = couchbase.cluster.PasswordAuthenticator(username, password)
+        self.cluster.authenticate(auth)
+        self.registry = self.cluster.open_bucket(bucket)
+        self.bucket = bucket
+
+    class RegistryUnavailable(Exception):
+        pass
+
+    def _get_associated_node_id(self, rtype, value):
+        if rtype == 'device':
+            return value['node_id']
+        elif rtype in ['receiver', 'sender', 'source', 'flow']:
+            return self.registry.get(value['device_id']).value['node_id']
+
+    def get_health(self, rkey, port=None):
+        return self.registry.lookup_in(
+            rkey, couchbase.subdocument.get('$document.exptime', xattr=True)
+        )['$document.exptime']
+
+    def upsert(self, rtype, rkey, value, xattrs, ttl=12):
+        try:
+            upsert_result = self.registry.upsert(rkey, value, ttl=ttl)
+            r = make_response(json.dumps(value), 200)
+        except couchbase.exceptions.KeyExistsError:
+            return make_response(409)
+        if upsert_result.success:
+            write_time = Timestamp.get_time().to_nanosec()
+            time.sleep(1)
+            subdoc_results = []
+
+            xattrs['last_updated'] = write_time
+            xattrs['created_at'] = write_time
+            xattrs['resource_type'] = rtype
+            xattrs['node_id'] = self._get_associated_node_id(rtype, value)
+
+            # Store any additional extended attributes
+            for key, value in xattrs.items():
+                subdoc_results.append(self.registry.mutate_in(
+                    rkey,
+                    couchbase.subdocument.upsert(_legacy_key_lookup(key), value, xattr=True)
+                ))
+
+            failed_subdoc_ops = [result for result in subdoc_results if result.success == False]
+            if len(failed_subdoc_ops) > 0:
+                return failed_subdoc_ops
+
+        if rtype != 'node':
+            ttl = self.get_health(xattrs['node_id'])
+
+        try:
+            touch_result = self.registry.touch(rkey, ttl=ttl)
+        except Exception:
+            self.remove(rkey)
+            r = make_response(500)
+
+        return r
+
+    # Legacy put command warps around upsert. Sanitises inputs, removing etcd specific decoration.
+    def put(self, rtype, rkey, value, ttl=12, port=None):
```

Comment:
The TTL default probably ought to come from config, and match the 'NODE_SEEN_TTL' from elsewhere.